crates/core/tedge_agent/src/restart_manager/actor.rs

```diff
@@ -50,16 +49,13 @@ impl Actor for RestartManagerActor {
         while let Some(request) = self.message_box.recv().await {
             let maybe_error = self.handle_restart_operation(&request).await;
 
-            match timeout(Duration::from_secs(5), self.message_box.recv_message()).await {
-                Ok(Some(WrappedInput::RuntimeRequest(RuntimeRequest::Shutdown))) => return Ok(()),
-                Ok(Some(WrappedInput::Message(second_request))) => {
-                    if let Err(err) = maybe_error {
-                        error!("{}", err);
-                    }
-                    self.handle_error(&request).await?;
-                    self.handle_error(&second_request).await?;
+            match timeout(Duration::from_secs(5), self.message_box.recv_signal()).await {
+                Ok(Some(RuntimeRequest::Shutdown)) => {
+                    // As expected, the restart triggered a shutdown.
+                    return Ok(());
                 }
-                _ => {
+                Ok(None) | Err(_) => {
+                    // Something went wrong. The process should have been shutdown by the restart.
                     if let Err(err) = maybe_error {
                         error!("{}", err);
                     }
```

Comment:
Comment for the one line below (GH doesn't show it). It's a question.

```rust
self.handle_error(&request).await?;
```
This is the only possibility to return `Err` in `run()` method. 
Since it returns `Ok` for all other errors, I am not sure if we should return `Err` for this? Maybe it should print error message and not return `Err`?

Reply:
So there are two errors here.

- The first one is caught by `let maybe_error = self.handle_restart_operation(&request).await;`
- The second one is the `Err(_)` case of the `match timeout(..)` call.

The second error is simply the timeout: no messages have been received during 5 seconds.
So we can assume that the restart fails. In that case we send the first error to the user:
this error why the restart operation failed (e.g. the user running the agent has not been granted authorization to reboot the device).

Reply:
After off-line discussion it appeared that an important issue were still lurking behind the question mark in `self.handle_error(&request).await?;` The event loop was failing to report these errors to the user and was stopping prematurely.

Fixed by https://github.com/thin-edge/thin-edge.io/pull/1968/commits/26daf2a126775f9b4d4fc17d53504605c2699fea